### DISCUSSION TITLE
External handling for PPK and PPK_ID

### DISCUSSION THREAD

**Discussion by Woll7** [strongswan/strongswan#243] [Q&A]
Hi,

Is it possible to externalize PPK and PPK_ID handling via a custom plugin?

The scenario here is for Site-to-Site gateways.

I'd like to be able (through a plugin or otherwise) to retrieve a random PPK_ID and PKK from an external service for the initiator, and the responder would look up the PPK_ID on it's end to retrieve the associated PPK (again via an external keying service).

In accordance with RFC 8784, the responder would look up the PPK from the sent PPK_ID, and if the keys are mismatched, the connection fails (this works now in Strongswan). I understand that NULL authentication (RFC 7619) is not implemented in Strongswan, but that's not an issue in this case.

I've looked into how to build plugins, so any pointers are appreciated.

Thanks in advance.

P.S. I cross posted this to the Devs list

**Comment by tobiasbrunner** [comment]
The PPK_ID (together with the `ppk_required` constraint) is part of connection config (`peer_cfg_t`). So that can't directly be provided/modified by a plugin unless it also provides the rest of the config (loading configs via vici at runtime makes this a bit more flexible than configuring via text files).

PPKs can be provided by plugins (via `credential_set_t` interface) as they are treated like any other shared secret (type is `SHARED_PPK`).

I wonder though, what you intend to use as key distribution service because downloading PPKs seems unsafe unless that channel is quantum safe too.
